main/src/library/Result.flix

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright 2016 Liam Palmer
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Result {
+
+    ///
+    /// Returns `true` iff `r` is `Ok(v)`.
+    ///
+    def isOk[t, e](r: Result[t, e]): Bool = match r with {
+        case Ok(v) => true
+        case Err(w) => false
+    }
+
+    ///
+    /// Returns `true` iff `r` is `Err(w)`.
+    ///
+    def isErr[t, e](r: Result[t, e]): Bool = match r with {
+        case Ok(v) => false
+        case Err(w) => true
+    }
+
+    ///
+    /// Returns `v` if `r` is `Ok(v)`. Undefined semantics if `r` is `Err(w)`.
+    ///
+    def get[t, e](r: Result[t, e]): t = match r with {
+        case Ok(v) => v
+    }
+
+    ///
+    /// Returns `v` if `r` is `Ok(v)`. Otherwise returns `d`.
+    ///
+    def getWithDefault[t, e](r: Result[t, e], d: t): t = match r with {
+        case Ok(v) => v
+        case Err(w) => d
+    }
+
+    ///
+    /// Returns `r1` if it is `Ok(v)`. Otherwise returns `r2`.
+    ///
+    def withDefault[t, e1, e2](r1: Result[t, e1], r2: Result[t, e2]): Result[t, e2] = match r1 with {
+        case Ok(v) => r1
+        case Err(w) => r2
+    }
+
+    ///
+    /// Returns `true` if `r` is `Ok(v)` and the predicate `f(v)` evaluates to `true`. Otherwise returns `false`.
+    ///
+    def exists[t, e](f: t -> Bool, r: Result[t, e]): Bool = getWithDefault(map(f, r), false)
+
+    ///
+    /// Returns `true` if `r` is `Ok(v)` and the predicate `f(v)` evaluates to `true` or if `r` is `Err(w)`.
+    /// Otherwise returns `false`.
+    ///
+    def forall[t, e](f: t -> Bool, r: Result[t, e]): Bool = getWithDefault(map(f, r), true)
+
+    ///
+    /// Returns `r` if `r` is `Ok(v)` and the predicate `f(v)` evaluates to true. Otherwise returns `Err(d)`.
+    ///
+    def filter[t, e](f: t -> Bool, r: Result[t, e], d: e): Result[t, e] = match r with {
```

Comment:
Delete. Filtering doesn't make sense for Result because there is no default empty value like `None`.